src/garage/envs/environment.py

```diff
@@ -0,0 +1,189 @@
+                of arrays corresponding to `action_spec()`.
+
+        Returns:
+            garage.TimeStep: The TimeStep after the action. Attributes:
+                * env_spec (garage.envs.EnvSpec): Environment specification.
+                * observation: an `np.ndarray`, or a nested dict, list or
+                    tuple corresponding to :obj:`env_spec.observation_space`.
+                    The observation before applying the action.
+                * action (np.ndarray): A NumPy array, or a nested dict, list or
+                    tuple of arrays corresponding to `action_spec()`.
+                * reward (float): The reward for taking the action given the
+                    observation, at the this time step.
+                * next_observation: an `np.ndarray`, or a nested dict, list or
+                    tuple corresponding to :obj:`env_spec.observation_space`.
+                    The observation after applying the action.
+                * terminal (bool): The termination signal for the this time
+                    step.
+                * env_info (dict): A dict arbitrary environment state
+                    information.
+                * agent_info (numpy.ndarray): A dict of arbitrary agent
+                    state information.
+                * step_type (garage.StepType): a `StepType` enum value. Can
+                    either be StepType.FIRST, StepType.MID, StepType.LAST,
+                    or StepType.TIMEOUT.
+
+        """
+
+    @abc.abstractmethod
+    def render(self, mode):
+        """Renders the environment.
+
+        The set of supported modes varies per environment. By convention,
+        if mode is:
+
+        * rgb_array: Return an numpy.ndarray with shape (x, y, 3),
+          representing RGB values for an x-by-y pixel image, suitable
+          for turning into a video.
+        * ansi: Return a string (str) or StringIO.StringIO containing a
+          terminal-style text representation. The text can include newlines
+          and ANSI escape sequences (e.g. for colors).
+
+        Make sure that your class's metadata 'render.modes' key includes
+        the list of supported modes.
+
+        Args:
+            mode (str): the mode to render with. The string must be present in
+            self.metadata['render.modes'].
+
+        Raises:
+            NotImplementedError: If the mode is not implemented.
+
+        For example:
+
+        class MyEnv(Env):
+            metadata = {'render.modes': ['display', 'rgb_array']}
+
+            def render(self, mode):
+                if mode == 'rgb_array':
+                    return np.array(...) # return RGB frame suitable for video
+                elif mode == 'ansi':
+                    ... # return text output
+                else:
+                    super(MyEnv, self).render(mode=mode) # just raise an
+                    exception
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def visualize(self):
```

Comment:
do i call this once or may times?

Reply:
I think we can call this once, and let reset() handles update of the visualization. 

Copied workflow from the doc string:
```
env.reset() # reset the environment
env.visualize() # open a window for visualization
act_space = env.action_space
action = act_space.sample() # sample a random action
env.step(action) # step the environment
# the display should be updated as well
env.close() # close the environment and the visualization
```

Reply:
I think a minimal example of the rollout loop like this one would be extremely helpful in the docstrings here. It is the most powerful way of showing people how your API is meant to be used.

Reply:
for instance, you can look at the metaworld README: https://github.com/rlworkgroup/metaworld